platform/ios/src/MGLScaleBarView.m

```diff
@@ -0,0 +1,289 @@
+    for (NSArray<NSNumber *> *row in rows) {
+        CLLocationDistance currentDiff = ABS(preferredDistance - row.firstObject.floatValue);
+        if (currentDiff < diff) {
+            preferredRow = row;
+        }
+    }
+    return preferredRow;
+}
+
+- (void)fadeIn {
+    if (self.isVisible) {
+        return;
+    }
+
+    self.visible = YES;
+    
+    [UIView animateWithDuration:.2 delay:0 options:UIViewAnimationOptionBeginFromCurrentState animations:^{
+        self.alpha = 1.0;
+    } completion:nil];
+}
+
+- (void)fadeOut {
+    if (!self.isVisible) {
+        return;
+    }
+    
+    self.visible = NO;
+    
+    [UIView animateWithDuration:.3 delay:0.7 options:UIViewAnimationOptionBeginFromCurrentState animations:^{
+        self.alpha = 0.0;
+    } completion:nil];
+}
+
+#pragma mark - Setters
+
+- (void)setMetersPerPoint:(CLLocationDistance)metersPerPoint {
+    if (_metersPerPoint == metersPerPoint) {
+        return;
+    }
+    
+    [self fadeIn];
+    
+    _metersPerPoint = metersPerPoint;
+    self.row = [self preferredRowInRows:[self validRows]];
+    
+    [self invalidateIntrinsicContentSize];
+    [self setNeedsLayout];
+}
+
+- (void)setRow:(NSArray<NSNumber *> *)row {
+    if (![row isEqualToArray:_row]) {
+        [_bars makeObjectsPerformSelector:@selector(removeFromSuperview)];
+        [_labels makeObjectsPerformSelector:@selector(removeFromSuperview)];
+        _bars = nil;
+        _labels = nil;
+        _row = row;
+    }
+}
+
+#pragma mark - Views
+
+- (NSArray<UIView *> *)bars {
+    if (!_bars) {
+        NSMutableArray *bars = [NSMutableArray array];
+        for (NSUInteger i = 0; i < self.row.lastObject.integerValue; i++) {
+            UIView *bar = [[UIView alloc] init];
+            [bars addObject:bar];
+            [self.containerView addSubview:bar];
+        }
+        _bars = bars;
+    }
+    return _bars;
+}
+
+- (NSArray<UILabel *> *)labels {
+    if (!_labels) {
+        NSMutableArray *labels = [NSMutableArray array];
+        for (NSUInteger i = 0; i <= self.row.lastObject.integerValue; i++) {
+            UILabel *label = [[UILabel alloc] init];
+            label.text = @"0";
+            label.font = [UIFont systemFontOfSize:8];
+            [label sizeToFit];
+            [labels addObject:label];
+            [self addSubview:label];
+        }
+        _labels = labels;
+    }
+    return _labels;
+}
+
+#pragma mark - Layout
+
+- (void)layoutSubviews {
+    [super layoutSubviews];
+    
+    if (!self.row.lastObject || !self.row.lastObject.integerValue) {
+        // Current distance is not within allowed range
+        return;
+    }
+    
+    [self updateLabels];
+    [self layoutBars];
+    [self layoutLabels];
+}
+
+- (void)layoutBars {
+    
+    CGFloat barWidth = floorf(self.bounds.size.width / self.bars.count);
+    
+    [self.bars enumerateObjectsUsingBlock:^(UIView * _Nonnull bar, NSUInteger idx, BOOL * _Nonnull stop) {
```

Comment:
Since you aren’t making use of functional programming, an ordinary for-in loop should be fine here.